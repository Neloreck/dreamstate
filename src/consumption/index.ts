export * from "./useManager";
export * from "./Consume";
export * from "./withConsumption";
