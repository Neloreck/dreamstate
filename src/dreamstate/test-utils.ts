/*
 * 'https://github.com/Neloreck/dreamstate'
 */

export { ScopeContext as InternalScopeContext } from "@/dreamstate/core/scoping/ScopeContext";

export * from "@/dreamstate/test-utils/registry/getCurrent";
export * from "@/dreamstate/test-utils/utils/getReactConsumer";
export * from "@/dreamstate/test-utils/utils/getReactProvider";

export * from "@/dreamstate/test-utils/registry/mockManagerWithScope";
export * from "@/dreamstate/test-utils/registry/mockScopeProvider";
export * from "@/dreamstate/test-utils/registry/mockScope";

export * from "@/dreamstate/test-utils/utils/nextAsyncQueue";
