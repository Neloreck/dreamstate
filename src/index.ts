import * as React from "react";
import {
  createContext,
  createElement,
  ComponentType,
  Context,
  Consumer,
  useContext,
  useLayoutEffect,
  useState, useRef, ReactElement
} from "react";

const IS_PRODUCTION: boolean = (process.env.NODE_ENV === "production");

/**
 * Utility.
 * Shallow comparison for objects.
 *
 * todo: use package? Not sure.
 */
export const shallowEqualObjects = (first: object, second: object): boolean => {

  if (first === second) {
    return true;
  }

  if (!first || !second) {
    return false;
  }

  const firstKeys: Array<string> = Object.keys(first);
  const secondKeys: Array<string> = Object.keys(second);

  const length: number = firstKeys.length;

  if (secondKeys.length !== length) {
    return false;
  }

  for (let it = 0; it < length; it ++) {

    const key: string = firstKeys[it];

    // @ts-ignore indexed for built-ins.
    if (first[key] !== second[key]) {
      return false;
    }
  }

  return true;
};

/*
 * Â±1.4KB min+gzip, only pure react codebase usage.
 *
 * 'https://github.com/Neloreck/dreamstate'
 *
 * OOP style context store for react.
 * Reactivity, lifecycle, strict TS typing, dependency injection + singleton pattern for each application store without boilerplate code.
 */

/**
 * Interface for string indexed objects.
 */
interface IStringIndexed<T> {
  [index: string]: T;
}

/**
 * Any context manager type.
 */
export type TAnyCM = ContextManager<any>;

/**
 * Pick declaration.
 * Describe object that will help to get listed state params.
 */
export interface IConsumePick<A extends TAnyCM> {
  from: A;
  take: Array<keyof A["context"]>;
}

/**
 * Generic setter method.
 */
export type TSetter<T> = (value: T) => void;

/**
 * Consumable type.
 * Manager or pick descriptor.
 */
export type TConsumable<T extends TAnyCM> = IConsumePick<T> | T;

/**
 * Use manager hook, higher order wrapper for useContext.
 */
export const useManager = <T extends object>(manager: ContextManager<T>): T => useContext(manager.internalReactContext);

/**
 * Generate subtree of providers.
 */
const generateProviderTree = (targets: Array<TAnyCM>, current: number, children: ReactElement): ReactElement => {
  return current < 0 ? children : createElement(targets[current].getProvider(), null, generateProviderTree(targets, current - 1, children));
};

/**
 * Decorator factory.
 * Provide context from context manager.
 * Observes changes and uses default react Provider for data flow.
 */
export const Provide = (...sources: Array<TAnyCM>): ClassDecorator => <T>(target: T) => {

  function P(props: IStringIndexed<any>) {
    return generateProviderTree(sources, sources.length - 1, createElement(target as any, props));
  }

  /**
   * Use correct naming for non-production mode.
   */
  if (IS_PRODUCTION) {
    P.displayName = "dp";
  } else {
    P.displayName = `Dreamstate.Provider[${sources.map((it: TConsumable<any>) => it.constructor.name )}]`;
  }

  return P as any;
};

/**
 * todo: Wait for variadic arguments from typescript.
 * Declaration interface (temp) for consumer decorator.
 */
export interface IConsume {
  // Mock for variadic selectors.
  <A extends TAnyCM, B extends TAnyCM, C extends TAnyCM, D extends TAnyCM, E extends TAnyCM, F extends TAnyCM>(
    a: TConsumable<A>, b?: TConsumable<B>, c?: TConsumable<C>, d?: TConsumable<D>, e?: TConsumable<E>, f?: TConsumable<F>,
  ): ClassDecorator;
  // Default usage with context managers.
  (...managers: Array<ContextManager<any>>): ClassDecorator;
}

/**
 * Decorator factory.
 * Consumes context from context manager.
 * Observes changes and uses default react Provider.
 */
export const Consume: IConsume =
  (...sources: Array<TConsumable<any>>): any => {

    return <C>(target: C) => {

      function C(props: any) {

        let consumed: IStringIndexed<any> = { ...props };

        for (const source of sources) {

          if (source instanceof ContextManager) {
            consumed = Object.assign(consumed, useManager(source))
          } else {

            const propsToPick: Array<string> = (source as IConsumePick<any>).take as Array<string>;
            const propsPicked: IStringIndexed<any> = useManager((source as IConsumePick<any>).from);

            consumed = Object.assign(consumed, propsToPick.reduce((a: IStringIndexed<any>, e: string) => (a[e] = propsPicked[e], a), {}));
          }
        }

        return createElement(target as any, { ...props, ...consumed });
      }

      /**
       * Use correct naming for non-production mode.
       */
      if (IS_PRODUCTION) {
        C.displayName = "dc";
      } else {
        C.displayName = `Dreamstate.Consumer[${sources.map((it: TConsumable<any>) => it instanceof ContextManager ?  it.constructor.name : `${it.from.constructor.name}{${it.take}}`)}]`;
      }

      return C;
    }
  };

/**
 * Abstract class.
 * Class based context manager for react.
 * Current Issue: Static items inside of each class instance.
 */
export abstract class ContextManager<T extends object> {

  /**
   * Setter method factory.
   * !Strictly typed generic method with 'update' lifecycle.
   * Helps to avoid boilerplate code with manual 'update' transactional updates for simple methods.
   */
  public static getSetter = <S extends object, D extends keyof S>(manager: ContextManager<S>, key: D) => {

    return (obj: Partial<S[D]>) => {
      manager.beforeUpdate();
      manager.context[key] = { ...(manager.context as any)[key], ...(obj as object) };
      manager.observedElements.forEach((it) => it(manager.getProvidedProps()));
      manager.afterUpdate();
    };
  };

  /**
   * Observer factory for react providers.
   * Allows to use lifecycle and observer pattern.
   *
   * todo: Investigate, if possible: Single observer for every @Provide element.
   */
  private static getObserver = <T extends IStringIndexed<any>>(parent: ContextManager<T>): ComponentType => {

    // todo: Stricter caching and prevent double-copy for ref.
    function Observer(props: IStringIndexed<any>) {

      const [ observedState, setObservedState ]: [ any, any ] = useState(parent.getProvidedProps());

      const stateRef = useRef(observedState);

      const setWithMemo = (newState: IStringIndexed<any>) => {

        if (Object.keys(stateRef.current).some((key: string): boolean => !shallowEqualObjects(stateRef.current[key], newState[key]))) {
          stateRef.current = newState;
          setObservedState(newState);
        }
      };

      useLayoutEffect(() => {

        if (parent.observedElements.length === 0) {
          parent.onProvisionStarted();
        }

        parent.observedElements.push(setWithMemo);

        return () => {

          parent.observedElements = parent.observedElements.filter((it) => it !== setWithMemo);

          if (parent.observedElements.length === 0) {
            parent.onProvisionEnded();
          }
        }
      }, []);

      return React.createElement(parent.internalReactContext.Provider,  { value: observedState }, props.children);
    }

    if (IS_PRODUCTION) {
      Observer.displayName = "dp";
    } else {
      Observer.displayName = `Dreamstate.Observer[${parent.constructor.name}]`;
    }

    return Observer;
  };

  /**
   * React Context<T> store internal.
   */
  public readonly internalReactContext: Context<T>;

  /**
   * Abstract store/actions bundle.
   * Left for generic implementation.
   */
  public abstract context: T;

  /**
   * Array of provider observers.
   * Used for one app-level supply or for separate react dom sub-trees injection.
   */
  protected observedElements: Array<TSetter<any>> = [];

  /**
   * Default constructor.
   * Allows to create react context provider/consumer bundle.
   * Stores it as private readonly singleton per each storage.
   */
  public constructor() {
    this.internalReactContext = createContext(this.getProvidedProps());
  }

  /**
   * Utility getter.
   * Allows to get related React.Consumer for manual renders.
   */
  public getConsumer(): Consumer<T> {
    return this.internalReactContext.Consumer;
  }

  /**
   * Utility getter.
   * Allows to get related React.Provider for manual renders.
   */
  public getProvider(): ComponentType {
    return ContextManager.getObserver(this);
  }

  /**
   * Force React.Provider update.
   * Calls lifecycle methods.
   * Should not cause odd renders because affects only related React.Provider elements (commonly - 1 per store).
   */
  public update(): void {

    this.beforeUpdate();
    this.observedElements.forEach((it: TSetter<any>) => it(this.getProvidedProps()));
    this.afterUpdate();
  }

  // todo: Set context?

  /**
   * Lifecycle.
   * First provider was injected into DOM / Last provider was removed from DOM.
   */
  protected onProvisionStarted(): void {}
  protected onProvisionEnded(): void {}

  /**
   * Lifecycle.
   * Before/after manual update lifecycle event.
   * Also shares for 'getSetter' methods.
   */
  protected beforeUpdate(): void {}
  protected afterUpdate(): void {}

  /**
   * Get provided context object.
   * Spread object every time for new references and provider HOC update.
   * It will not force consumers/React.Provider odd renders because actions/state nested objects will be separated.
   */
  private getProvidedProps(): T {
    return Object.assign({}, this.context);
  }

}

/**
 * Decorator factory.
 * Modifies method descriptor, so it will be bound to prototype instance once.
 * All credits: 'https://www.npmjs.com/package/autobind-decorator'.
 */
export const Bind = (): MethodDecorator => <T>(target: object, propertyKey: string | symbol, descriptor: TypedPropertyDescriptor<T>): TypedPropertyDescriptor<T> => {

  let original: T = descriptor.value as T;

  if (typeof original !== "function") {
    throw new TypeError(`@Bind decorator can only be applied to methods not: ${typeof original}.`);
  }

  // In IE11 calling Object.defineProperty has a side-effect of evaluating the
  // getter for the property which is being replaced. This causes infinite
  // recursion and an "Out of stack space" error.
  let definingProperty = false;

  return {
    configurable: true,
    get(): T {

      // @ts-ignore functional interface.
      if (definingProperty || this === target.prototype || this.hasOwnProperty(propertyKey) || typeof original !== 'function') {
        return original;
      }

      const bound: T = original.bind(this);

      definingProperty = true;

      Object.defineProperty(this, propertyKey, {
        configurable: true,
        get(): T {
          return bound;
        },
        set(value: T): void {
          original = value;
          delete this[propertyKey];
        }
      });

      definingProperty = false;

      return bound;
    },
    set(value: T): void {
      original = value;
    }
  };
};
