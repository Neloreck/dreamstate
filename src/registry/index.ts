export * from "./addServiceObserverToRegistry";
export * from "./getCurrentContext";
export * from "./getCurrent";
export * from "./registerService";
export * from "./removeServiceObserverFromRegistry";
export * from "./subscribeToManager";
export * from "./unRegisterService";
export * from "./unSubscribeFromManager";
export * from "./getReactContext";
