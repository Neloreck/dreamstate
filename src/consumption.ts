import {
  ComponentType,
  createElement,
  MutableRefObject,
  useCallback, useContext,
  useLayoutEffect,
  useRef,
  useState
} from "react";
import { default as hoistNonReactStatics } from "hoist-non-react-statics";

import {
  ClassDescriptor,
  IConsume,
  IConsumeDecorator,
  IContextManagerConstructor,
  IStringIndexed,
  TConsumable,
  TTakeContextSelector,
  TUpdateSubscriber
} from "./types";
import { EMPTY_ARR, EMPTY_STRING, IDENTIFIER_KEY, MANAGER_REGEX } from "./internals";
import { ContextManager } from "./ContextManager";
import { addManagerSubscriber, removeManagerSubscriber, STORE_REGISTRY } from "./registry";

declare const IS_DEV: boolean;

/**
 * Use manager hook with subscribed updates.
 * Same like common useContext hook, but has memo checks.
 */
export function useContextWithMemo<T extends object, D extends IContextManagerConstructor<T>>(
  managerConstructor: D,
  depsSelector: (context: T) => Array<any>
): D["prototype"]["context"] {
  const [ state, setState ] = useState(function() {
    return STORE_REGISTRY.CONTEXT_STATES[managerConstructor[IDENTIFIER_KEY]];
  });
  const observed: MutableRefObject<Array<any>> = useRef(depsSelector(state));

  const updateMemoState: TUpdateSubscriber<T> = useCallback(function(nextContext: T): void {
    // Calculate changes like react lib does and fire change only after update.
    const nextObserved = depsSelector(nextContext);

    for (let it = 0; it < nextObserved.length; it ++) {
      if (observed.current[it] !== nextObserved[it]) {
        observed.current = nextObserved;
        setState(nextContext);
        return;
      }
    }
  }, EMPTY_ARR);

  useLayoutEffect(function() {
    addManagerSubscriber(managerConstructor, updateMemoState);
    return function() {
      removeManagerSubscriber(managerConstructor, updateMemoState);
    };
  });

  return state;
}

/**
 * Use manager hook, higher order wrapper for useContext.
 */
export function useManager<T extends object, D extends IContextManagerConstructor<T>>(
  managerConstructor: D,
  depsSelector?: (context: D["prototype"]["context"]) => Array<any>
): D["prototype"]["context"] {
  if (depsSelector) {
    return useContextWithMemo(managerConstructor, depsSelector);
  } else {
    return useContext(managerConstructor.getContextType());
  }
}

/**
 * Function for consume wrappers that maps selectors and allows class components to consume store data.
 */
export function createManagersConsumer(target: ComponentType, sources: Array<TConsumable<any>>) {
  // todo: Should we warn if no alias and no selector provided?
  // todo: Should we warn if provided empty array-selector?
  /**
   * Extended assistance for DEV bundle.
   */
  if (IS_DEV) {
    // Warn about too big consume count.
    if (sources.length > 5) {
      console.warn("Seems like your component tries to consume more than 5 stores at once, more than 5 can" +
        " lead to slower rendering for big components. Separate consuming by using multiple Consume decorators/hocs " +
        "for one component or review your components structure.",
      `Source: '${target.name}'.`
      );
    }

    // Validate input sources.
    for (const source of sources) {
      if (typeof source === "object") {
        if (!source.from || typeof source.from !== "function") {
          throw new TypeError(
            "Specified 'from' selector should point to correct context manager. Supplied type: " +
            `'${typeof source.from}'. Check '${target.name}' component.`
          );
        } else if (!(source.from.prototype instanceof ContextManager)) {
          throw new TypeError(
            "Specified consume target should inherit 'ContextManager', seems like you have forgotten to extend your " +
            `manager class. Wrong parameter: '${source.from.name || "anonymous."}'. Check '${target.name}' component.`);
        }

        if (typeof source.as !== "undefined" && typeof source.as !== "string") {
          throw new TypeError(
            "Specified 'as' param should point to string component property key. Supplied type: " +
            `'${typeof source.as}'. Check '${target.name}' component.`
          );
        }

        if (source.take === null) {
          throw new TypeError(
            "Specified 'take' param should be a valid selector. Selectors can be functional, string or array. " +
            `Supplied type: '${typeof source}'. Check '${target.name}' component.`
          );
        }
      } else if (typeof source === "function") {
        if (!(source.prototype instanceof ContextManager)) {
          throw new TypeError(
            "Specified consume target should inherit 'ContextManager', seems like you forgot to extend your manager " +
            `class. Wrong parameter: '${source.name || "anonymous function"}'. Check '${target.name}' component.`);
        }
      } else {
        throw new TypeError(
          "Specified consume source is not selector or context manager." +
          `Supplied type: '${typeof source}'. Check '${target.name}' component.`
        );
      }
    }
  } else {
    for (const source of sources) {
      if (
        (typeof source !== "object" && typeof source !== "function")
        ||
        (
          typeof source === "object" && (
            (!source.from || typeof source.from !== "function" || !(source.from.prototype instanceof ContextManager)) ||
            (typeof source.as !== "undefined" && typeof source.as !== "string") ||
            source.take === null
          )
        )
        ||
        typeof source === "function" && !(source.prototype instanceof ContextManager)
      ) {
        throw new TypeError("Wrong context-consume parameter supplied.");
      }
    }
  }

  /**
   * Create selectors once to prevent checks on every render.
   * Optimization on resolving + additional checks.
   */
  const selectors: Array<(accumulator: IStringIndexed<any>) => void> = new Array(sources.length);

  for (let it = 0; it < selectors.length; it ++) {
    const source: TConsumable<any> = sources[it];

    // Is context manager.
    if (source.prototype instanceof ContextManager) {
      selectors[it] = (accumulator: IStringIndexed<any>) => Object.assign(accumulator, useManager(source));
    } else {
      const take: TTakeContextSelector<any> = source.take;
      const alias: string | undefined = source.as;

      // No selector, only alias.
      if (take === undefined) {
        if (alias) {
          selectors[it] = function(accumulator: IStringIndexed<any>) {
            accumulator[source.as] = useManager(source.from);
          };
        } else {
          selectors[it] = function(accumulator: IStringIndexed<any>) {
            Object.assign(accumulator, useManager(source.from));
          };
        }
      } else if (Array.isArray(take)) {
        // Selected array of needed props, filter only needed and alias it if 'as' is supplied.
        // Here we can automatically use memo parameter.
        const memoCheck = function(current: IStringIndexed<any>) {
          return (take as Array<string>).map(function(it: string) {
            return current[it];
          });
        };

        if (alias) {
          selectors[it] = function(accumulator: IStringIndexed<any>) {
            const context: IStringIndexed<any> = useManager(source.from, memoCheck);

            accumulator[alias] = (take as Array<string>).reduce((a: IStringIndexed<any>, e: string) =>
              (a[e] = context[e], a), {});
          };
        } else {
          selectors[it] = function(accumulator: IStringIndexed<any>) {
            const context: IStringIndexed<any> = useManager(source.from, memoCheck);

            Object.assign(accumulator, (take as Array<string>).reduce(function(a: IStringIndexed<any>, e: string) {
              return (a[e] = context[e], a);
            }, {}));
          };
        }
      } else if (typeof take === "function") {
        // Supplied functional selector, return object with needed props like redux does. Alias if 'as' is supplied.
        const memoCheck = function(current: IStringIndexed<any>) {
          return Object.values(take(current));
        };

        if (alias) {
          selectors[it] = function(accumulator: IStringIndexed<any>) {
            accumulator[alias] = take(useManager(source.from, memoCheck));
          };
        } else {
          selectors[it] = function(accumulator: IStringIndexed<any>) {
            Object.assign(accumulator, take(useManager(source.from, memoCheck)));
          };
        }
      } else if (typeof take === "string") {
        // Pick only selected key prop. Using memo-selector here.
        const memoCheck = function(current: IStringIndexed<any>) {
          return [ current[take] ];
        };

        if (alias) {
          selectors[it] = function(accumulator: IStringIndexed<any>) {
            accumulator[alias] = useManager(source.from, memoCheck)[take];
          };
        } else {
          selectors[it] = function(accumulator: IStringIndexed<any>) {
            accumulator[take] = useManager(source.from, memoCheck)[take] ;
          };
        }
      }
    }
  }

  // HOC component to pick props and provide needed/selected.
  function Consumer(ownProps: object) {
    const consumed: IStringIndexed<any> = {};

    // todo: Array.reduce or for-of?
    for (const selector of selectors) {
      selector(consumed);
    }

    return createElement(target as any, Object.assign(consumed, ownProps));
  }

  if (IS_DEV) {
    Consumer.displayName = `Dreamstate.Consumer.[${sources.map((it: TConsumable<any>) =>
      it.prototype instanceof ContextManager
        ? it.name.replace(MANAGER_REGEX, EMPTY_STRING)
        : `${it.from.name.replace(MANAGER_REGEX, EMPTY_STRING)}{${it.take}}`)}]`;
  } else {
    Consumer.displayName = "DS.Consumer";
  }

  return Consumer;
}

/**
 * Decorator factory.
 * Consumes context from context manager.
 * Observes changes and uses default react Provider.
 */
export const Consume: IConsumeDecorator = function(...sources: Array<TConsumable<any>>): any {
  // Higher order decorator to reserve params.
  return function(classOrDescriptor: ComponentType) {
    // Support legacy and proposal decorators.
    return ((typeof classOrDescriptor === "function"))
      ? hoistNonReactStatics(createManagersConsumer(classOrDescriptor, sources), classOrDescriptor)
      : ({
        ...(classOrDescriptor as ClassDescriptor),
        finisher: (wrappedComponent: ComponentType) =>
          hoistNonReactStatics(createManagersConsumer(wrappedComponent, sources), wrappedComponent)
      });
  };
} as any;

/**
 * HOC alias for @Consume.
 */
export const withConsumption: IConsume = Consume as IConsume;
