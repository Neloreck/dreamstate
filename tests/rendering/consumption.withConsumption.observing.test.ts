import { ComponentType, createElement } from "react";
import { mount, ReactWrapper } from "enzyme";
import { act } from "react-dom/test-utils";

import { withConsumption } from "../../src/consumption";
import { getCurrentManager } from "../../src/registry";

import {
  ITestContext,
  RenderCallbacker,
  TestContextManager,
  TextContextManagerProvider
} from "../assets";
import { nextAsyncQueue, registerManagerClass, unRegisterManagerClass } from "../helpers";

describe("HoC @Consume and withConsumption selector observing.", () => {
  function mountProvided<T extends object>(element: ComponentType<T>, props: T): ReactWrapper {
    return mount(createElement(TextContextManagerProvider, {}, createElement(element, props)));
  }

  beforeEach(() => {
    registerManagerClass(TestContextManager);
  });

  afterEach(() => {
    unRegisterManagerClass(TestContextManager);
  });

  it("Should properly observe only one property and prevent odd updates.", async () => {
    const manager: TestContextManager = getCurrentManager(TestContextManager)!;
    const Observer = withConsumption([ { from: TestContextManager, take: "first" } ])(RenderCallbacker);
    const onRender = jest.fn();

    const tree = mountProvided(Observer, { onRender });

    expect(tree).toMatchSnapshot();
    expect(onRender).toBeCalledTimes(1);

    onRender.mockClear();

    await act(async () => {
      manager.setContext({ first: "anotherValue" });
      await nextAsyncQueue();
    });

    tree.update();

    expect(tree).toMatchSnapshot();
    expect(onRender).toBeCalledTimes(1);

    onRender.mockClear();

    /**
     * Set same state again.
     */
    await act(async () => {
      manager.setContext({ first: "anotherValue" });
      await nextAsyncQueue();
    });

    tree.update();

    expect(tree).toMatchSnapshot();
    expect(onRender).not.toBeCalled();

    onRender.mockClear();

    /**
     * Set not observed context property.
     */
    await act(async () => {
      manager.setContext({ second: 9000 });
      await nextAsyncQueue();
    });

    tree.update();

    expect(tree).toMatchSnapshot();
    expect(onRender).not.toBeCalled();
  });

  it("Should properly observe array selected props and prevent odd updates.", async () => {
    const manager: TestContextManager = getCurrentManager(TestContextManager)!;
    const Observer = withConsumption([ { from: TestContextManager, take: [ "first", "third" ] } ])(RenderCallbacker);
    const onRender = jest.fn();

    const tree = mountProvided(Observer, { onRender });

    expect(tree).toMatchSnapshot();
    expect(onRender).toBeCalledTimes(1);

    onRender.mockClear();

    await act(async () => {
      manager.setContext({ first: "anotherValue" });
      await nextAsyncQueue();
    });

    tree.update();

    expect(tree).toMatchSnapshot();
    expect(onRender).toBeCalledTimes(1);

    onRender.mockClear();

    /**
     * Set same state again.
     */
    await act(async () => {
      manager.setContext({ first: "anotherValue" });
      await nextAsyncQueue();
    });

    tree.update();

    expect(tree).toMatchSnapshot();
    expect(onRender).not.toBeCalled();

    onRender.mockClear();

    /**
     * Set not observed context property.
     */
    await act(async () => {
      manager.setContext({ second: 9000 });
      await nextAsyncQueue();
    });

    tree.update();

    expect(tree).toMatchSnapshot();
    expect(onRender).not.toBeCalled();

    onRender.mockClear();

    /**
     * Set additional array-observed property.
     */
    await act(async () => {
      manager.setContext({ third: true });
      await nextAsyncQueue();
    });

    tree.update();

    expect(tree).toMatchSnapshot();
    expect(onRender).toBeCalledTimes(1);

    onRender.mockClear();
  });

  it("Should properly observe selector picked props and prevent odd updates.", async () => {
    const manager: TestContextManager = getCurrentManager(TestContextManager)!;
    const Observer = withConsumption([
      { from: TestContextManager, take: (context: ITestContext) => ({ prop: context.second }) }
    ])(RenderCallbacker);
    const onRender = jest.fn();

    const tree = mountProvided(Observer, { onRender });

    expect(tree).toMatchSnapshot();
    expect(onRender).toBeCalledTimes(1);

    onRender.mockClear();

    await act(async () => {
      manager.setContext({ first: "anotherValue" });
      await nextAsyncQueue();
    });

    tree.update();

    expect(tree).toMatchSnapshot();
    expect(onRender).not.toBeCalled();

    onRender.mockClear();

    await act(async () => {
      manager.setContext({ second: 56656 });
      await nextAsyncQueue();
    });

    tree.update();

    expect(tree).toMatchSnapshot();
    expect(onRender).toBeCalled();

    onRender.mockClear();
  });
});
