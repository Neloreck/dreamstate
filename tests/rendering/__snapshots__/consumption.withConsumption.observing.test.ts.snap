// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HoC @Consume and withConsumption selector observing. Should properly observe array selected props and prevent odd updates. 1`] = `
<Memo(DS.Observer)>
  <DS.Consumer
    onRender={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "first": "first",
              "third": false,
            },
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
  >
    <RenderCallbacker
      first="first"
      onRender={
        [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "first": "first",
                "third": false,
              },
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        }
      }
      third={false}
    >
      <div>
        {"first":"first","third":false}
      </div>
    </RenderCallbacker>
  </DS.Consumer>
</Memo(DS.Observer)>
`;

exports[`HoC @Consume and withConsumption selector observing. Should properly observe array selected props and prevent odd updates. 2`] = `
<Memo(DS.Observer)>
  <DS.Consumer
    onRender={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "first": "anotherValue",
              "third": false,
            },
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
  >
    <RenderCallbacker
      first="anotherValue"
      onRender={
        [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "first": "anotherValue",
                "third": false,
              },
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        }
      }
      third={false}
    >
      <div>
        {"first":"anotherValue","third":false}
      </div>
    </RenderCallbacker>
  </DS.Consumer>
</Memo(DS.Observer)>
`;

exports[`HoC @Consume and withConsumption selector observing. Should properly observe array selected props and prevent odd updates. 3`] = `
<Memo(DS.Observer)>
  <DS.Consumer
    onRender={[MockFunction]}
  >
    <RenderCallbacker
      first="anotherValue"
      onRender={[MockFunction]}
      third={false}
    >
      <div>
        {"first":"anotherValue","third":false}
      </div>
    </RenderCallbacker>
  </DS.Consumer>
</Memo(DS.Observer)>
`;

exports[`HoC @Consume and withConsumption selector observing. Should properly observe array selected props and prevent odd updates. 4`] = `
<Memo(DS.Observer)>
  <DS.Consumer
    onRender={[MockFunction]}
  >
    <RenderCallbacker
      first="anotherValue"
      onRender={[MockFunction]}
      third={false}
    >
      <div>
        {"first":"anotherValue","third":false}
      </div>
    </RenderCallbacker>
  </DS.Consumer>
</Memo(DS.Observer)>
`;

exports[`HoC @Consume and withConsumption selector observing. Should properly observe array selected props and prevent odd updates. 5`] = `
<Memo(DS.Observer)>
  <DS.Consumer
    onRender={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "first": "anotherValue",
              "third": true,
            },
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
  >
    <RenderCallbacker
      first="anotherValue"
      onRender={
        [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "first": "anotherValue",
                "third": true,
              },
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        }
      }
      third={true}
    >
      <div>
        {"first":"anotherValue","third":true}
      </div>
    </RenderCallbacker>
  </DS.Consumer>
</Memo(DS.Observer)>
`;

exports[`HoC @Consume and withConsumption selector observing. Should properly observe only one property and prevent odd updates. 1`] = `
<Memo(DS.Observer)>
  <DS.Consumer
    onRender={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "first": "first",
            },
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
  >
    <RenderCallbacker
      first="first"
      onRender={
        [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "first": "first",
              },
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        }
      }
    >
      <div>
        {"first":"first"}
      </div>
    </RenderCallbacker>
  </DS.Consumer>
</Memo(DS.Observer)>
`;

exports[`HoC @Consume and withConsumption selector observing. Should properly observe only one property and prevent odd updates. 2`] = `
<Memo(DS.Observer)>
  <DS.Consumer
    onRender={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "first": "anotherValue",
            },
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
  >
    <RenderCallbacker
      first="anotherValue"
      onRender={
        [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "first": "anotherValue",
              },
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        }
      }
    >
      <div>
        {"first":"anotherValue"}
      </div>
    </RenderCallbacker>
  </DS.Consumer>
</Memo(DS.Observer)>
`;

exports[`HoC @Consume and withConsumption selector observing. Should properly observe only one property and prevent odd updates. 3`] = `
<Memo(DS.Observer)>
  <DS.Consumer
    onRender={[MockFunction]}
  >
    <RenderCallbacker
      first="anotherValue"
      onRender={[MockFunction]}
    >
      <div>
        {"first":"anotherValue"}
      </div>
    </RenderCallbacker>
  </DS.Consumer>
</Memo(DS.Observer)>
`;

exports[`HoC @Consume and withConsumption selector observing. Should properly observe only one property and prevent odd updates. 4`] = `
<Memo(DS.Observer)>
  <DS.Consumer
    onRender={[MockFunction]}
  >
    <RenderCallbacker
      first="anotherValue"
      onRender={[MockFunction]}
    >
      <div>
        {"first":"anotherValue"}
      </div>
    </RenderCallbacker>
  </DS.Consumer>
</Memo(DS.Observer)>
`;

exports[`HoC @Consume and withConsumption selector observing. Should properly observe selector picked props and prevent odd updates. 1`] = `
<Memo(DS.Observer)>
  <DS.Consumer
    onRender={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "prop": 2,
            },
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
  >
    <RenderCallbacker
      onRender={
        [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "prop": 2,
              },
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        }
      }
      prop={2}
    >
      <div>
        {"prop":2}
      </div>
    </RenderCallbacker>
  </DS.Consumer>
</Memo(DS.Observer)>
`;

exports[`HoC @Consume and withConsumption selector observing. Should properly observe selector picked props and prevent odd updates. 2`] = `
<Memo(DS.Observer)>
  <DS.Consumer
    onRender={[MockFunction]}
  >
    <RenderCallbacker
      onRender={[MockFunction]}
      prop={2}
    >
      <div>
        {"prop":2}
      </div>
    </RenderCallbacker>
  </DS.Consumer>
</Memo(DS.Observer)>
`;

exports[`HoC @Consume and withConsumption selector observing. Should properly observe selector picked props and prevent odd updates. 3`] = `
<Memo(DS.Observer)>
  <DS.Consumer
    onRender={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "prop": 56656,
            },
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
  >
    <RenderCallbacker
      onRender={
        [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "prop": 56656,
              },
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        }
      }
      prop={56656}
    >
      <div>
        {"prop":56656}
      </div>
    </RenderCallbacker>
  </DS.Consumer>
</Memo(DS.Observer)>
`;
